pkg/utils/Idpool.go

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (c) 2022-2023 Intel Corporation, or its subsidiaries.
+// Copyright (C) 2023 Nordix Foundation.
+
+// Package linuxgeneralmodule is the main package of the application
+
+package utils
+
+import (
+	"log"
+	"reflect"
+	// "github.com/opiproject/opi-evpn-bridge/pkg/vendor_plugins/intel-e2000/p4runtime/p4translation"
+)
+
+// IDPool structure
+/*  Helper class for uniquely assigning IDs from a specified integer set (e.g. a
+#   range) to keys. IDs are assigned (or read) with get_id(key) and returned back
+#   into the pool with release_id(key). The IDPool remembers a once-assigned ID
+#   for keys so that the same ID is assigned for a key. Only when the pool runs
+#   out of unassigned keys, it will recycle released ids and assign them to new
+#   keys.
+#   Optionally, the IDPool supports reference tracking for key/ID pairs. Clients
+#   can provide a unique reference when fetching and releasing an ID for a key
+#   to support multiple independent clients.
+#   The pool will only release the ID for the key, when the last client has the
+#   released the ID with its reference. When a reference is specified in get_id()
+#   and release_id() the IDPool returns the current number of reference for the
+#   ID so that a caller knows when an ID was newly assigned (ref_count 1) or
+#   finally released (ref_count 0).
+# ---------------------------------------------------------------------------------
+*/
+type IDPool struct {
+	// self._lock = threading.Lock()
+	name         string                 // Name of pool
+	_unusedIDs   []uint32               // Yet unused IDs in pool Available ids
+	_idsInUse    map[interface{}]uint32 // Mapping key: id for currently assigned ids
+	_idsForReuse map[interface{}]uint32 // Mapping key: id for previously assigned ids
+	_refs        map[uint32][]interface{}
```

Comment:
refs sounds like a counter tracking how many times we requested the same id, no?

Reply:
correct, its reference to the same id

Reply:
then why do we need a slice of slices then? Can we have a slice of counters? `map[uint32]uint32`?

Reply:
actually we need golang type equivalent of 'list' type in python.
we have replaced with "ref map[uint32]map[interface{}]bool"
here the scenario is for same key we can have different references of interface type as reference can be of any type like int, string, alphanumeric etc. that's why we cannot use counter here.